implement OpenMetrics support for real

vroldanbet opened the pull request:
I discovered that thanos-federate-proxy lies about its capabilities: it's not truly OpenMetrics compliant.

# Context

The prometheus metrics scrape endpoint representation differs from OpenMetrics in subtle ways. Consumers who only support OpenMetrics fail to scrape anything produced by `thanos-federate-proxy`. We've observed that various scrappers in the wild are having difficulties parsing what's returned by `thanos-federate-proxy`.

For context, `thanos-federate-proxy` retrieves from the upstream using Prometheus Query API. This API returns `model.Value`, a wire representation that does not have all the same information presented in a scrape endpoint (`MetricsFamily`). So as it stands, this project can not reconstruct an OpenMetrics compliant representation with fundamental changes.

# How

If we transform the `model.Value` to `MetricsFamily` _somehow_, we could use the `expfmt` package from the `prometheus/common` library with the official bits for content negotiation and serialization in the scrape endpoint textual representation. This would mean the generated prometheus/openmetrics endpoint representation will be as accurate and as good what the prometheus project does.

The challenge here was how to transform `model.Value` to `MetricsFamily`. I learned that Prometheus offers `Metadata API`, which gives us the missing bits. Unfortunately, Thanos does not expose this API, so this commit requires you to wire the metadata API through the upstream prometheus. API Queries will hit Thanos, and metadata API calls will hit prometheus. The latter is done outside the critical path on an interval currently hardcoded to 1 minute. There could be a scenario where Metadata for a new metric is not yet present (as the async worker hasn't refreshed the metadata); in that case, the metric will be set to `unknown` type.

I don't know if the transformation is done correctly, but it delivers what we need. Please proceed with caution; this is experimental.

# What's included

- supports content negotiation from upstream `prometheus/common` using the `Accept` header
- returns `Content-Type` with the negotiated metrics representation, which helps with scrappers like the Datadog Agent
- reconstructs type information using Prometheus Metadata API, thus introducing a dependency on a prometheus instance. This includes `TYPE` and `HELP` comments

# Caveats

There are some gaps in this implementation, which is a stopgap that will buy us time to move on to something else. This project is not in good shape; it has barely been tested, and its design choices are quite naive. For that reason, it's not worth investing further time in it, and I recommend moving on to a more sustainable long-term option.

- Thanos could collect metrics from multiple prometheus clusters, but this implementation only supports one upstream Metadata API endpoint. In scenarios with multiple independent prometheus clusters, this implementation will fail to retrieve Metadata from the various prometheus clusters. This could be addressed in a follow-up.
- This could break at scale: retrieving large Metadata API responses, the implications of keeping all that metadata in memory, having a lock to access it, cloning it on each federate request, the size of the generated `/federate` endpoint, etc...
- Tests are not added for the new behavior (not any existed for the critical path of the project)

vroldanbet committed:
implement OpenMetrics support for real

I discovered that thanos-federate-proxy lies about its capabilities:
it's not truly OpenMetrics compliant.

The prometheus metrics scrape endpoint representation differs from
OpenMetrics in subtle ways. Consumers who only support OpenMetrics
fail to scrape anything produced by `thanos-federate-proxy`.

For context, `thanos-federate-proxy` retrieves from the upstream
using Prometheus Query API. This API returns `model.Value`, a wire
representation that does not have all the same information presented
in a scrape endpoint (`MetricsFamily`). If we transform it to `MetricsFamily`
_somehow_, we could use the `expfmt` package from `prometheus/common`
library with the official bits for content negotiation and serialization
in the scrape endpoint textual representation.

So that's what this PR does.

The challenge here was how to transform `model.Value` to `MetricsFamily`.
I discovered Prometheus offers `Metadata API`, which gives us the missing bits.
Unfortunately, Thanos does not expose this API, so this commit requires you
to wire the metadata API through the upstream prometheus. API Queries will
hit Thanos, and metadata API calls will hit prometheus. The latter is done
outside the critical path on an interval currently hardcoded to 1 minute.
There could be a scenario where Metadata for a new metric is not yet
present (as the async worker hasn't refreshed the metadata) and in that
case the metric will be set to `unknown` type.

I don't know if the transformation is done correctly, but so far, it
looks to deliver what we need. Please be sure to proceed with caution,
this is experimental.

vroldanbet committed:
return Content-Type header

this is used by some scrapers (like the DD prometheus
integration), otherwise they fail to parse the response.

e.g. DD prometheus integration expects the Content-Type
header to be set to either application/vnd.google.protobuf
or text/plain

vroldanbet committed:
fix type information returned

all #TYPE headers had "unknown" because
we were using the metric name without the suffix.

vroldanbet committed:
fix types not found in metadata registry

some metrics are registered in the registry
with the suffix included, so we don't need to attempt
to strip them out. We address this by checking the type
information with the unstripped name.

vroldanbet committed:
add missing #EOF in OpenMetrics

by calling the encoder.Close method

vroldanbet committed:
regression: metrics families actually need the stripped name

otherwise you end up with double suffixes like `_bucket_bucket`
in the openmetrics text represention

alecmerdler commented on main.go:
```diff
@@ -84,6 +105,35 @@ func main() {
+			mutex.Unlock()
+			if err != nil {
+				klog.Errorf("error refreshing metric metadata: %s", err.Error())
+				time.Sleep(250 * time.Millisecond)
```
This seems like a good candidate for using `time.Ticker`.

alecmerdler commented on main.go:
```diff
@@ -84,6 +105,35 @@ func main() {
+				klog.Errorf("error refreshing metric metadata: %s", err.Error())
+				time.Sleep(250 * time.Millisecond)
+			} else {
+				time.Sleep(1 * time.Minute)
```
This seems like a good candidate for using `time.Ticker`.

vroldanbet replied:
fixed in https://github.com/authzed/thanos-federate-proxy/pull/1/commits/46d82bf651e58acaf37ac15a728a60b54a6c971a

alecmerdler reviewed (approved):
Minor comment, otherwise looks good. Could we potentially try to upstream these changes so we don't need to rely on another fork?

vroldanbet committed:
PR feedback: make metadata worker loop use a ticker

vroldanbet commented:
>Could we potentially try to upstream these changes so we don't need to rely on another fork?

@alecmerdler, I'm not sure; this is a stopgap. I'm not sure this is the right strategy for this project, and now this introduces a dependency on prometheus, plus thanos can handle multiple prometheuses and this PR does not support that. From my perspective, this is exclusively to buy us time.

vroldanbet committed:
overhaul metric generation

The MetricFamily values were not being generated correctly:
- histograms need Buckets clustered for the same set of labels,
  this is done using the label fingerprints, which prometheus provides
  a helper for
- since __names__ is part of the labelset, and that typically comes
  with the suffix, we can't cluster together _sum and _count because
  they end up with a different fingerprint. We fix this by changing
  the name of to the stripped value. We still need the name to disambiguate
  in case there are other histograms with the same labelset
- "le" also need removed from the labelset, as it's generated when
  rendered
- clustering for gauges and counts is not necessary
- spec indicates that buckets need to be ordered by "le", so
  a sort call is added per histogram. See
  https://github.com/prometheus/OpenMetrics/blob/main/specification/OpenMetrics.md#histogram-1

>Buckets MUST be sorted in number increasing order of "le", and the value of the "le" label MUST follow the rules for Canonical Numbers.

- Ordering by metric family is optional by the spec
- Histogram counts and sum are properly set now
